Encrypt configs at rest

jkitman opened the pull request:
Since the guardian servers will be publicly known we should reduce the attack surface by not storing private keys in plaintext which would be easy targets for attackers with access to the filesystem or a backup of the config file.  This PR encrypts the server configs containing private keys at-rest using a password supplied by the user.

- Password can be supplied at a prompt during config generation and server start-up to avoid being in command history
- Uses PBKDF2_HMAC_SHA256 to hash the password to reduce the effectiveness of brute-force attacks
- Uses CHACHA20_POLY1305 to authenticate and encrypt the configs (so they cannot be tampered with)

Note we will need to modify the nonce generation if we intend to generate multiple versions of the configs with the same password (once we have a need for versioned configs).

Fix #724

elsirion commented on fedimintd/src/lib.rs:
```diff
@@ -0,0 +1,61 @@
+    LessSafeKey::new(key)
+}
+
+pub fn tls_nonce() -> Nonce {
+    Nonce::assume_unique_for_key([0; NONCE_LEN])
+}
+
+// FIXME - encrypting different versions of the configs will require different nonces to be secure
+pub fn config_nonce() -> Nonce {
+    Nonce::assume_unique_for_key([1; NONCE_LEN])
+}
```
Is "nonce" here a different term for initialization vector? Hard coded nonces are a big red flag imo. Can we deterministically derive it from the key/content somehow? Is it used for en/decryption or just authentication?

elsirion replied:
Ok, do you have a plan for fixing it? Password re-use would trigger the same, right? So I'd like to fix this ASAP if we land it like that.

elsirion replied:
> In the case we need to generate new versions of the configs we could have an auto-incrementing counter stored in the config file.

But if the config is encrypted we wouldn't know the counter at decryption time, right? If we don't need to know it then we should just hash the config and use that as the nonce.

jkitman replied:
We could store the incrementing nonce in the config file itself as plaintext.  I could make that change now I suppose.

elsirion commented on modules/fedimint-wallet/src/config.rs:
```diff
@@ -175,7 +175,7 @@ impl WalletClientConfig {
             peg_in_descriptor,
             network: Network::Regtest,
-            finality_delay: 0,
+            finality_delay: FINALITY_DELAY,
```
Good catch!

jkitman commented on fedimintd/src/lib.rs:
```diff
@@ -0,0 +1,61 @@
+    let password = match password {
+        None => rpassword::prompt_password("Enter a password to encrypt configs: ").unwrap(),
+        Some(password) => {
+            println!("WARNING: Passing in a password from the command line may be less secure!");
```
For output in our main files it seems like we prefer to use `println!`, not sure why but we should probably have a separate PR to change them all at once.

jkitman commented on fedimintd/src/bin/distributedgen.rs:
```diff
@@ -136,9 +152,9 @@ async fn run_dkg(
-    let pk_string = fs::read_to_string(dir_out_path.join(TLS_PK)).expect("Can't read file.");
     let cert_string = fs::read_to_string(dir_out_path.join(TLS_CERT)).expect("Can't read file.");
-    let pk = rustls::PrivateKey(hex::decode(pk_string).expect("not hex encoded"));
+    let pk = rustls::PrivateKey(encrypted_read(key, tls_nonce(), dir_out_path.join(TLS_PK)));
```
What do you mean by native encryption?

elsirion replied:
Maybe I'm wrong, but I thought there was a native way to en/decrypt PEM files. Right now it's kind of an ad-hoc scheme.

jkitman commented:
> I think we should get better at justifying our design decisions when it comes to cryptographic protocols (what I'm trying to do with #811). This PR isn't too complex, so it won't be a lot of work and could even be added as a doc string for e.g. a module that contains the en/decryption functionality. What do you think?

I added some docs that should explain the crypto choices: https://github.com/jkitman/minimint/blob/encrypt-configs-at-rest/fedimintd/src/lib.rs#L1-L15

Also changed the nonces from hard-coded to auto-incrementing and stored in the file, this will support generating new config versions in the future.

elsirion commented on fedimintd/src/bin/distributedgen.rs:
```diff
@@ -88,29 +95,37 @@ async fn main() {
-            println!("{}", gen_tls(&dir_out_path, address, base_port, name));
+            let salt: [u8; 16] = rand::random();
+            fs::write(dir_out_path.join(SALT_FILE), hex::encode(salt)).expect("write error");
+            let key = get_key(password, dir_out_path.join(SALT_FILE));
```
~Eventually we can probably devise a more compact file format (one binary encoded file instead of two), but for now this seems good.~

Ah, that's a different salt for the password. Actually more of a pepper since it's unique to the user and kinda private.

elsirion commented on fedimintd/src/lib.rs:
```diff
@@ -0,0 +1,93 @@
+
+    /// Reads encrypted data from a file, returns an incremented nonce for encrypting the next file
+    pub fn encrypted_read(key: &LessSafeKey, file: PathBuf) -> (Vec<u8>, Nonce) {
+        let hex = fs::read_to_string(file).expect("Can't read file.");
```
Not sure why hex, but not a blocker imo.
